sunpy/timeseries/timeseriesbase.py

```diff
@@ -376,28 +378,55 @@ def concatenate(self, otherts, same_source=False, **kwargs):
         Notes
         -----
         Extra keywords are passed to `pandas.concat`.
+
+        Examples
+        --------
+        A single `~sunpy.timeseries.TimeSeries` or an `collections.abc.Iterable` containing multiple
+        `~sunpy.timeseries.TimeSeries` can be passed to concatenate.
+
+        >>> timeseries_1.concatenate(timeseries_2) # doctest: +SKIP
+        >>> timeseries_1.concatenate([timeseries_2, timeseries_3]) # doctest: +SKIP
+
+        Set `same_source` to `True` if the sources of the time series are the same.
+
+        >>> timeseries_1.concatenate([timeseries_2, timeseries_3], same_source=True) # doctest: +SKIP
         """
-        # TODO: decide if we want to be able to concatenate multiple time series at once.
-        # check to see if nothing needs to be done
-        if self == otherts:
+        # Check to see if nothing needs to be done in case the same TimeSeries is provided.
+        if self == others:
             return self
+        elif isinstance(others, Iterable):
+            if len(others) == 1 and self == next(iter(others)):
+                return self
 
         # Check the sources match if specified.
-        if same_source and not (isinstance(otherts, self.__class__)):
-            raise TypeError("TimeSeries classes must match if specified.")
-
-        # Concatenate the metadata and data
-        kwargs['sort'] = kwargs.pop('sort', False)
-        meta = self.meta.concatenate(otherts.meta)
-        data = pd.concat([self._data.copy(), otherts.to_dataframe()], **kwargs)
+        if same_source:
+            if isinstance(others, Iterable):
+                if not all(isinstance(series, self.__class__) for series in others):
+                    raise TypeError("TimeSeries classes must match if 'same_source' is specified.")
+            else:
+                if not isinstance(others, self.__class__):
+                    raise TypeError("TimeSeries classes must match if 'same_source' is specified.")
```

Comment:
I feel like this might be refactorable to a certain extent since both error messages are the same.

The reason we check if it's an iterable first is that if it's not, the call to `all()` throws a TypeError which stops execution. Another way of doing it might be to use a try-except, but even that isn't significantly different from what this already does.
Something along the lines of -

```python
if same_source:
    try:
        if not all(isinstance(series, self.__class__) for series in others):
            raise TypeError("...")
    except TypeError:
        if not instance(others, self.__class__):
            raise TypeError("...")
```

Reply:
Would an `or` not work?

Reply:
Yeah, I tried that! So say we write the following -

```python
if same_source:
    if (
        isinstance(others, Iterable)
        and not all(isinstance(series, self.__class__) for series in others)
    ) or not isinstance(others, self.__class__):
        raise TypeError("TimeSeries classes must match if 'same_source' is specified.")
```

and have `goes_1` and `goes_data = [goes_2, goes_3, goes_4]` and you try the following - `goes_1.concatenate(goes_data, same_source=True)`, it will throw an error (even though it shouldn't).

What's happening is -

`isinstance(others, Iterable)` = True (which we have in place to short-circuit the first check if an iterable isn't passed) and `not all(isinstance(series, self.__class__) for series in others)` = False, so the first check becomes False. It goes on to the second check where `isinstance(others, self.__class__)` = True because `others` is a list. 

So to solve it we need to add something to short-circuit the second check if an iterable is passed to it which is basically `not isinstance(others, Iterable)`.

If that's what we're writing then we might as well make it more readable with cleaner if-statements.

Something tells me that I might be missing something blatantly obvious though, but I just can't figure out what 😓

Reply:
Maybe a chain of if/elif/else statements makes more sense.

Reply:
Something like this?

```python
if (
    same_source
    and isinstance(others, Iterable)
    and not all(isinstance(series, self.__class__) for series in others)
):
    raise TypeError("TimeSeries classes must match if 'same_source' is specified.")
elif (
    same_source
    and not isinstance(others, Iterable)
    and not isinstance(others, self.__class__)
):
    raise TypeError("TimeSeries classes must match if 'same_source' is specified.")
```

Reply:
Sure